src/mpid/ch4/shm/xpmem/xpmem_pre.h

```diff
@@ -36,16 +58,26 @@ typedef struct MPIDI_XPMEM_segtree {
 typedef struct {
     xpmem_segid_t remote_segid;
     xpmem_apid_t apid;
-    MPIDI_XPMEM_segtree_t segcache;     /* AVL tree based segment cache */
+    MPIDI_XPMEM_segtree_t segcache_user;        /* AVL tree based segment cache for user buffer */
+    MPIDI_XPMEM_segtree_t segcache_cnt; /* AVL tree based segment cache for XPMEM counter buffer */
 } MPIDI_XPMEM_segmap_t;
 
+typedef struct MPIDI_XPMEM_cnt_free_obj {
+    MPIDI_XPMEM_cnt_t *counter_ptr;
+    struct MPIDI_XPMEM_cnt_free_obj *prev, *next;
+} MPIDI_XPMEM_cnt_free_obj_t;
+
 typedef struct {
     int num_local;
     int local_rank;
     MPIR_Group *node_group_ptr; /* cache node group, used at win_create. */
     xpmem_segid_t segid;        /* my local segid associated with entire address space */
     MPIDI_XPMEM_segmap_t *segmaps;      /* remote seg info for every local processes. */
     size_t sys_page_sz;
+    MPIDI_XPMEM_seg_t **coop_counter_seg_direct;        /* original direct cooperative counter array vitural address,
+                                                         * used for detach in finalize routine */
+    MPIDI_XPMEM_cnt_t **coop_counter_direct;    /* direct cooperative counter array attached in xpmem initialization */
+    int num_pending_cnt;
```

Comment:
Can we compress "in xpmem initialization" to "at init"? I feel many of your comments are over-detailed.